server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java

```diff
@@ -290,14 +331,24 @@ public ClusterMetadataManifest writeIncrementalMetadata(
             previousClusterState.metadata(),
             clusterState.metadata()
         ) == false;
-        String globalMetadataFile;
-        // For migration case from codec V0 to V1, we have added null check on global metadata file,
-        // If file is empty and codec is 1 then write global metadata.
-        if (updateGlobalMetadata || previousManifest.getGlobalMetadataFileName() == null) {
-            globalMetadataFile = writeGlobalMetadata(clusterState);
-        } else {
-            logger.debug("Global metadata has not updated in cluster state, skipping upload of it");
-            globalMetadataFile = previousManifest.getGlobalMetadataFileName();
+
+        final boolean updateCoordinationMetadata = Metadata.isCoordinationMetadataEqual(
+            previousClusterState.metadata(),
+            clusterState.metadata()
+        ) == false;
+        final boolean updateSettingsMetadata = Metadata.isSettingsMetadataEqual(
+            previousClusterState.metadata(),
+            clusterState.metadata()
+        ) == false;
+        final boolean updateTemplatesMetadata = Metadata.isTemplatesMetadataEqual(
+            previousClusterState.metadata(),
+            clusterState.metadata()
+        ) == false;
+        final Map<String, UploadedMetadataAttribute> previousStateCustomMap = new HashMap<>(previousManifest.getCustomMetadataMap());
+        final Map<String, Metadata.Custom> customsToUpload = getUpdatedCustoms(clusterState, previousClusterState);
+        final Map<String, UploadedMetadataAttribute> allUploadedCustomMap = new HashMap<>(previousManifest.getCustomMetadataMap());
+        for (final String custom : clusterState.metadata().customs().keySet()) {
+            previousStateCustomMap.remove(custom);
```

Comment:
the whole logic around generating the latest uploaded customs map is very complex as of now, can we simplify it?

you can iterate over the latest customs metadata in clusterState and populate two maps which can be used later. Here is a rough pseudo code

```
allUploadedCustomMap = new Map
toUploadMap = new Map
previousCustomsMap = previousClusterState.metadata().customs().keySet();
for (String custom: clusterState.metadata().customs().keySet()) {
      if (previousCustomsMap.contains(custom) && isCustomsEqual(clusterState.metadata().customs().get(custom), previousClusterState.metadata().customs().get(custom)) {
              allUploadedCustomMap.put(custom, previousManifest.getRemoteFileName(custom));
       } else {
              toUploadMap.put(custom, clusterState.metadata().customs().get(custom))
       }
}

allUploadedFiles = writetoToRemote(toUploadMap)
allUploadedCustomMap.putAll(allUploadedFiles.customsMetadataFilesMap)
```

finally `allUploadedCustomMap` will contain the custom metadata files for the latest manifest.

Reply:
I am finding updated Customs same way as we were doing for IndexMetadata. But agreed both IndexMetadata and Custom Metadata update calculation is little tricky and can be better by the codeflow you have given above.